File: src/test/java/integration/com/company/crm/rules/RemoveDuplicateThreadsRuleTest.java
package com.company.crm.rules;

import com.company.crm.entity.Company;
import com.company.crm.entity.Conversation;
import com.company.crm.entity.Thread;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;

@ContextConfiguration(classes = {RemoveDuplicateThreadsRule.class})
@ExtendWith(SpringExtension.class)
class RemoveDuplicateThreadsRuleTest {

    @Autowired
    private RemoveDuplicateThreadsRule removeDuplicateThreadsRule;

    /**
     * given a valid company list with conversations containing duplicate conversations, the duplicate conversations should be removed.
     * given a valid company list with conversations containing non-duplicate threads, the list should be returned as it.
     * given a valid company list with no conversations , the same company list should be returned as is.
     */

    @Test
    public void testRemoveDuplicateThreadAndPopulateDuplicateThreadsCollection() {
        //given
        Thread thread_one = Thread.builder()
                .id(123L)
                .payload("some identical payload")
                .build();
        Thread thread_two = Thread.builder()
                .id(345L)
                .payload("some identical payload")
                .build();

        Conversation conversation = Conversation.builder()
                .id(111L)
                .from("demo_one@mail.com")
                .received(LocalDateTime.now())
                .number(555L)
                .userId(777L)
                .threads(new HashSet<>(List.of(thread_one, thread_two)))
                .build();
        Company company = com.company.crm.entity.Company.builder()
                .id(1L)
                .name("companyXyz")
                .signedUp(LocalDateTime.now())
                .conversations(Set.of(conversation))
                .build();
        List<Company> companies = new ArrayList<>(List.of(company));

        removeDuplicateThreadsRule.execute(companies);
        assertThat(companies.get(0).getConversations().stream().findFirst().get().getDuplicateThreads().size()).isEqualTo(1);
        assertThat(companies.get(0).getConversations().stream().findFirst().get().getThreads().size()).isEqualTo(1);
    }

    @Test
    public void testNoDuplicateThreadsConversationListIsNotChanged() {
        //given
        Thread thread_one = Thread.builder()
                .id(123L)
                .payload("some identical payload")
                .build();
        Thread thread_two = Thread.builder()
                .id(345L)
                .payload("some non identical payload")
                .build();

        Conversation conversation = Conversation.builder()
                .id(111L)
                .from("demo_one@mail.com")
                .received(LocalDateTime.now())
                .number(555L)
                .userId(777L)
                .threads(new HashSet<>(List.of(thread_one, thread_two)))
                .build();
        Company company = com.company.crm.entity.Company.builder()
                .id(1L)
                .name("companyXyz")
                .signedUp(LocalDateTime.now())
                .conversations(Set.of(conversation))
                .build();
        List<Company> companies = new ArrayList<>(List.of(company));
        //when
        removeDuplicateThreadsRule.execute(companies);
        //then
        assertThat(companies.get(0).getConversations().stream().findFirst().get().getDuplicateThreads().size()).isEqualTo(0);
        assertThat(companies.get(0).getConversations().stream().findFirst().get().getThreads().size()).isEqualTo(2);
    }

    @Test
    public void testCompanyWithNoConversationsListIsNotChanged() {
        Company company = com.company.crm.entity.Company.builder()
                .id(1L)
                .name("companyXyz")
                .signedUp(LocalDateTime.now())
                .build();
        List<Company> companies = new ArrayList<>(List.of(company));
        removeDuplicateThreadsRule.execute(companies);
        assertThat(companies.get(0).getConversations().stream().findFirst().isEmpty()).isEqualTo(true);
        assertThat(companies.get(0).getConversations().stream().findFirst().isEmpty()).isEqualTo(true);
    }
}

File: src/test/java/com/company/crm/review/rules/RemoveDuplicateThreadsRuleTest.kt
package com.company.crm.review.rules

import com.company.crm.review.domain.CompanyDto
import com.company.crm.review.domain.CompanyDto.ConversationDto
import com.company.crm.review.domain.CompanyDto.ConversationDto.ThreadDto
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class RemoveDuplicateThreadsRuleTest {
    
    private val removeDuplicateThreadsRuleService = RemoveDuplicateThreadsRule()
    
    @Test
    fun `should remove duplicate threads`() {
        val companies = listOf(
            company(1,
                conversation(10,
                    thread(101, "a"),
                    thread(102, "a"),
                    thread(103, "b")
                )
            ),
            company(2,
                conversation(20,
                    thread(201, "a"),
                    thread(202, "b"),
                    thread(203, "b")
                )
            )
        )

        removeDuplicateThreadsRuleService.execute(companies)

        assertThat(companies)
                .flatExtracting("conversations")
                .flatExtracting("threads")
                .extracting("id")
                .containsExactly(101L, 103L, 201L, 202L)
    }

    private fun company(id: Long, vararg conversations: ConversationDto) = CompanyDto().apply {
        this.id = id
        this.name = "Company $id"
        this.conversations = listOf(*conversations)
    }

    private fun conversation(id: Long, vararg threads: ThreadDto) = ConversationDto().apply {
        this.id = id
        this.threads = listOf(*threads)
    }

    private fun thread(id: Long, payload: String) = ThreadDto().apply {
        this.id = id
        this.payload = payload
    }
}